src/browser/BrowserService.cpp

```diff
@@ -607,6 +614,177 @@ QString BrowserService::getKey(const QString& id)
+                nullptr, origin, rpId, rpName, username, publicKeyCredentials.id, userHandle, publicKeyCredentials.key);
+        }
+
+        hideWindow();
+        return publicKeyCredentials.response;
+    }
+
+    hideWindow();
+    return getPasskeyError(ERROR_PASSKEYS_REQUEST_CANCELED);
+}
+
+// Passkey authentication
+QJsonObject BrowserService::showPasskeysAuthenticationPrompt(const QJsonObject& publicKey,
+                                                             const QString& origin,
+                                                             const StringPairList& keyList)
+{
+    auto db = selectedDatabase();
+    if (!db) {
+        return getPasskeyError(ERROR_KEEPASS_DATABASE_NOT_OPENED);
+    }
+
+    const auto userVerification = publicKey["userVerification"].toString();
+    if (!browserPasskeys()->isUserVerificationValid(userVerification)) {
+        return getPasskeyError(ERROR_PASSKEYS_INVALID_USER_VERIFICATION);
+    }
+
+    // Parse "allowCredentials"
+    const auto rpId = publicKey["rpId"].toString();
+    const auto entries = getPasskeyAllowedEntries(publicKey, rpId, keyList);
+    if (entries.isEmpty()) {
+        return getPasskeyError(ERROR_KEEPASS_NO_LOGINS_FOUND);
+    }
+
+    // With single entry, if no verification is needed, return directly
+    if (entries.count() == 1 && userVerification == BrowserPasskeys::REQUIREMENT_DISCOURAGED) {
+        return getPublicKeyCredentialFromEntry(entries.first(), publicKey, origin);
+    }
+
+    const auto timeout = publicKey["timeout"].toInt();
+
+    raiseWindow();
+    BrowserPasskeysConfirmationDialog confirmDialog;
+    confirmDialog.authenticateCredential(entries, origin, timeout);
+    auto dialogResult = confirmDialog.exec();
+    if (dialogResult == QDialog::Accepted) {
+        hideWindow();
+        const auto selectedEntry = confirmDialog.getSelectedEntry();
+        return getPublicKeyCredentialFromEntry(selectedEntry, publicKey, origin);
+    }
+
+    hideWindow();
+    return getPasskeyError(ERROR_PASSKEYS_REQUEST_CANCELED);
+}
+
+void BrowserService::addPasskeyToGroup(Group* group,
+                                       const QString& url,
+                                       const QString& rpId,
+                                       const QString& rpName,
+                                       const QString& username,
+                                       const QString& userId,
+                                       const QString& userHandle,
+                                       const QString& privateKey)
+{
+    // If no group provided, use the default browser group of the selected database
+    if (!group) {
+        auto db = selectedDatabase();
+        if (!db) {
+            return;
+        }
+        group = getDefaultEntryGroup(db);
+    }
+
+    auto* entry = new Entry();
+    entry->setUuid(QUuid::createUuid());
+    entry->setGroup(group);
+    entry->setTitle(QString("%1 (%2)").arg(rpName, tr("Passkey")));
```

Comment:
```suggestion
    entry->setTitle(tr("%1 (Passkey)").arg(rpName));
```

Reply:
I actually liked the way this was coded originally, clever to not include the formatting in the translation if you can avoid it.

Reply:
It robs translators of the opportunity to change the order and perhaps there are different conventions for when and how you use parentheses in some languages. I would generally avoid making any assumptions about user-facing strings.

Reply:
I'll change those. RTL languages are a good example where those translations will break.

Reply:
Good points